### PR TITLE
Feature/hash pwkeys 20210401

### PR DESCRIPTION
**Pull request by cschu**
- addresses #238 
- password reset tokens are now hashed before they're stored in the database
- adapted password reset integration tests to hashed reset tokens
- security.get_new_password_reset_token can now generate tokens with arbitrary life time
- based on integration tests introduced in #257

### CODE REVIEW

File: datameta/security.py [hunk 1]
```diff
@@ -30,29 +30,30 @@
...
+def get_new_password_reset_token(db:Session, user:User, expires=datetime.now() + timedelta(minutes=10)):
     """Clears all password recovery tokens for user identified by the supplied
     email address, generates a new one and returns it.
 
     Returns:
-        None  - If the user is not found or disabled
-        token - otherwise"""
+        - PasswordToken object (with hashed token value)
+        - unhashed token value
+    """
```
Comment (lkuchenb) [inline datameta/security.py:40]:
```suggestion
    """
    expires = datetime.now() + timedelta(minutes=10) if expires is None else expires
```

File: tests/integration/test_pwupdate.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+"""
+import time
+from datetime import datetime, timedelta
+
+from . import BaseIntegrationTest
+from datameta.api import base_url
+from .utils import create_pwtoken
+
+
+class TestPasswordUpdate(BaseIntegrationTest):
```
Comment (lkuchenb) [inline tests/integration/test_pwupdate.py:11]: Can you look into using `parameterized` here, too?

**Review by lkuchenb — CHANGES_REQUESTED** [review]
Thank you for the patches and additions! Here are my remarks:

File: datameta/views/forgot.py [hunk 3]
```diff
@@ -51,12 +51,13 @@ def v_forgot_api(request):
     if not re.match(r"[^@]+@[^@]+\.[^@]+", req_email):
         return { 'success' : False, 'error' : 'MALFORMED_EMAIL' }
 
-    token = security.get_new_password_reset_token_from_email(db, req_email)
+    reset_token = security.get_new_password_reset_token_from_email(db, req_email)
```
[the remark this answers is not in the document]
Comment (cschu) [reply datameta/views/forgot.py:54]: yea agreed - should the `v_forgot_api` still return `success` in that case? I assume yes, for obfuscation reasons?

File: tests/integration/test_pwupdate.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+    def test_failure_expired_reset_token(self, status:int=410):
+        """Testing unsuccessful (self-)password change with expired token.
+
+        Expected Response:
+            HTTP 410
+        """
+        user = self.users["user_a"]
+        pwtoken = create_pwtoken(self.session_factory, user, expires=datetime.now() + timedelta(seconds=1))
+        time.sleep(5)
```
[the remark this answers is not in the document]
Comment (cschu) [reply tests/integration/test_pwupdate.py:122]: i did that initially, but it errored (i think it violates a constraint on the timestamp)

File: datameta/views/forgot.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lkuchenb) [reply datameta/views/forgot.py:54]: Yup

File: tests/integration/test_pwupdate.py (hunk 2 above)
Comment (cschu) [reply tests/integration/test_pwupdate.py:12]: 6/8 tests 'parameterized'

File: datameta/security.py (hunk 1 above)
Comment (cschu) [reply datameta/security.py:40]: done.

File: datameta/security.py [hunk 5]
```diff
@@ -30,29 +30,30 @@
 def generate_token():
     return "".join(choice(ascii_letters+digits) for _ in range(64) )
 
-def get_new_password_reset_token(db:Session, user:User):
+def get_new_password_reset_token(db:Session, user:User, expires=datetime.now() + timedelta(minutes=10)):
```
[the remark this answers is not in the document]
Comment (cschu) [reply datameta/security.py]: done.

File: tests/integration/test_pwupdate.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (cschu) [reply tests/integration/test_pwupdate.py]: works now.

File: datameta/views/forgot.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cschu) [reply datameta/views/forgot.py]: done.

File: datameta/views/forgot.py [hunk 6]
```diff
@@ -51,12 +51,13 @@ def v_forgot_api(request):
...
 
     # Generate a new forgot token and send it to the user
     if token:
-        send_forgot_token(request, token)
-        log.debug(f"USER REQUESTED RECOVERY TOKEN: {token.value}")
+        db_token_obj, clear_token = reset_token
```
[the remark this answers is not in the document]
Comment (cschu) [reply datameta/views/forgot.py]: done.

**Review by lkuchenb — APPROVED** [review]
Thanks, I've added a last style commit, good to merge from my end!

**Pull request merged**